internal/agent/list.go

```diff
@@ -12,41 +12,86 @@ import (
+	agentPerPage     int
+	agentLastPage    int
+	agentsLoading    bool
+	agentLoader      func(int, int) tea.Cmd
 }
 
-func NewAgentList(loader tea.Cmd) AgentListModel {
+func NewAgentList(loader func(int, int) tea.Cmd, page, perpage int) AgentListModel {
 	l := list.New(nil, NewDelegate(), 0, 0)
 	l.Title = "Buildkite Agents"
 
 	return AgentListModel{
-		agentList:   l,
-		agentLoader: loader,
+		agentList:        l,
+		agentCurrentPage: page,
+		agentPerPage:     perpage,
+		agentLoader:      loader,
 	}
 }
 
+func (m *AgentListModel) appendAgents() tea.Cmd {
+	// Set agentsLoading
+	m.agentsLoading = true
+	// Set a status message and start the agentList's spinner
+	startSpiner := m.agentList.StartSpinner()
+	setStatus := m.agentList.NewStatusMessage(("Fetching more agents"))
+	// Fetch and append more agents
+	appendAgents := m.agentLoader(m.agentCurrentPage, m.agentPerPage)
+	return tea.Sequence(appendAgents, tea.Batch(startSpiner, setStatus))
+}
+
 func (m AgentListModel) Init() tea.Cmd {
-	return m.agentLoader
+	return m.agentLoader(m.agentCurrentPage, m.agentPerPage)
 }
 
 func (m AgentListModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 	var cmds []tea.Cmd
 	switch msg := msg.(type) {
 	case tea.WindowSizeMsg:
-		// when viewport size is reported, start a spinner and show a message to the user indicating agents are loading
+		// When viewport size is reported, start a spinner and show a message to the user indicating agents are loading
 		h, v := agentListStyle.GetFrameSize()
 		m.agentList.SetSize(msg.Width-h, msg.Height-v)
 		return m, tea.Batch(m.agentList.StartSpinner(), m.agentList.NewStatusMessage("Loading agents"))
+	case tea.KeyMsg:
+		switch msg.String() {
+		case "down":
+			// Calculate last element, unfiltered and if the last agent page via the API has been reached
+			lastListItem := m.agentList.Index() == len(m.agentList.Items())-1
+			unfilteredState := m.agentList.FilterState() == list.Unfiltered
+			if !m.agentsLoading && lastListItem && unfilteredState {
+				lastPageReached := m.agentCurrentPage > m.agentLastPage
+				// If down is pressed on the last agent item, list state is unfiltered and more agents are available
+				// to load the API
+				if !lastPageReached {
+					return m, m.appendAgents()
+				} else {
+					// Append a status message to alert that no more agents are available to load from the API
+					setStatus := m.agentList.NewStatusMessage("No more agents to load!")
+					cmds = append(cmds, setStatus)
+				}
+			}
+		}
+	// Custom messages
 	case NewAgentItemsMsg:
-		// when a new page of agents is received, append them to existing agents in the list and stop the loading
+		// When a new page of agents is received, append them to existing agents in the list and stop the loading
 		// spinner
-		allItems := append(m.agentList.Items(), msg.Items()...)
+		allItems := append(m.agentList.Items(), msg.ListItems()...)
 		cmds = append(cmds, m.agentList.SetItems(allItems))
+		// Stop the loading spinner
```

Comment:
I dont think we need to comment every line, especially if its just the same as reading the line its documenting

Reply:
👍